pipeline/test_email_notifier.py

```diff
@@ -0,0 +1,160 @@
+"""Tests for email_notifier.py"""
+
+from unittest import mock
+from unittest.mock import MagicMock, patch
+import pytest
+from boto3.exceptions import Boto3Error
+from email_notifier import (
+    get_ses_client,
+    get_subscriptions_and_products,
+    send_email,
+    check_and_notify
+)
+
+
+def test_get_ses_client_missing_env_variables():
+    """Test if an exception is raised when required environment variables are missing."""
+    with mock.patch.dict("os.environ", {}, clear=True):
+        with pytest.raises(RuntimeError, match="Required environment variables for AWS or email are missing"):
+            get_ses_client()
+
+
+def test_get_ses_client_success():
+    """Test the successful creation of an SES client."""
+    with mock.patch.dict("os.environ", {
+        "AWS_ACCESS_KEY_ID": "fake_key_id",
+        "AWS_ACCESS_SECRET_KEY": "fake_secret_key",
+        "FROM_EMAIL": "test@example.com"
+    }), mock.patch("boto3.client") as mock_boto_client:
+
+        mock_ses_client = MagicMock()
+        mock_boto_client.return_value = mock_ses_client
+
+        ses_client = get_ses_client()
+        assert ses_client is mock_ses_client
+        mock_boto_client.assert_called_once_with(
+            "ses", region_name="eu-west-2", aws_access_key_id="fake_key_id",
+            aws_secret_access_key="fake_secret_key"
+        )
+
+
+def test_get_subscriptions_and_products():
+    """Test the subscription retrieval logic."""
+
+    mock_cursor = MagicMock()
+
+    mock_connection = MagicMock()
+
+    mock_cursor.fetchall.return_value = [
+        (1, 10.0, "Product A", "user@example.com")]
+
+    with patch("email_notifier.get_cursor") as mock_get_cursor:
+        mock_get_cursor.return_value.__enter__.return_value = mock_cursor
+
+        subscriptions = get_subscriptions_and_products(mock_connection)
+
+    assert len(subscriptions) == 1
+    assert subscriptions[0] == (1, 10.0, "Product A", "user@example.com")
+
+
+def test_send_email_success():
+    """Test sending an email using SES."""
+    mock_ses_client = MagicMock()
+    mock_ses_client.send_email.return_value = {
+        "ResponseMetadata": {"HTTPStatusCode": 200}
+    }
+
+    with mock.patch.dict("os.environ", {
+        "FROM_EMAIL": "test@example.com",
+        "AWS_ACCESS_KEY_ID": "fake_key_id",
+        "AWS_ACCESS_SECRET_KEY": "fake_secret_key"
```

Comment:
Optional: Minor - I think this is repeated a few times, could this be refactored into a `@pytest.fixure`? - I know theres still a few things to sort with the tests so this can be considered at the end!

Something like:
```
@pytest.fixture
def aws_test_credentials():
    with mock.patch.dict("os.environ", {
        "FROM_EMAIL": "test@example.com",
        "AWS_ACCESS_KEY_ID": "fake_key_id",
        "AWS_ACCESS_SECRET_KEY": "fake_secret_key"})

Reply:
Im not 100% this code I wrote above works by the way! Just a rough example!